### PR TITLE
Stream export

### PR DESCRIPTION
**Pull request by tpope**
This adds streaming support to exports, improving response time and greatly reducing memory usage. All 3 formats are supported, but note that XLSX still creates the entire spreadsheet in memory under the hood, so the gains there will be more limited. Addresses #5187, #9533, and #11463.

This started as an attempt to resolve conflicts on #9614, but after fixing bugs, cleaning up some design issues, and extending it to JSON and XLSX, I don't think much of the original PR remains. [A comment on that issue](https://github.com/metabase/metabase/pull/9614#pullrequestreview-281171595) stated changes of this nature wouldn't necessarily require tests; crossing my fingers that it still holds.

###### Before submitting the PR, please make sure you do the following
- [N/A] If you're attempting to fix a translation issue, please submit your changes to our [POEditor project](https://poeditor.com/join/project/ynjQmwSsGh) instead of opening a PR.
### Tests
-  [x] Run the frontend and integration tests with  `yarn lint && yarn flow && yarn test`)
-  [x] If there are changes to the backend codebase, run the backend tests with `lein test && lein lint && ./bin/reflection-linter`

-  [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).

### CODE REVIEW

File: src/metabase/driver/sql_jdbc/execute.clj
Comment (walterl) [inline src/metabase/driver/sql_jdbc/execute.clj]: Does this `if` deliberately have no else clause? If so, how about making it a `when`?

File: src/metabase/api/embed.clj
Comment (walterl) [inline src/metabase/api/embed.clj]: This line is getting a little long. Would you mind splitting it up into multiple lines, please?

File: src/metabase/query_processor.clj
Comment (walterl) [inline src/metabase/query_processor.clj]: It's a bit odd to me that the decoration of the results with the column info and row formatting, be done here. Would it not be more appropriate to do this higher up the call stack? Maybe it should even be defined as part of `f`, so that `execute-query` doesn't have to know about formatting. What do you think?

File: src/metabase/query_processor.clj
Comment (walterl) [inline src/metabase/query_processor.clj]: Is there a specific design reason why you opted to add `:data-fn` args, instead of popping it into `options` (which gets merged into `query`)? Is it to make it clearer, from the function signature, that it's different from the base case (async)?

File: src/metabase/query_processor/middleware/format_rows.clj
Comment (walterl) [inline src/metabase/query_processor/middleware/format_rows.clj]: Why did you decide to factor out this function? If you wanted to make sure you get a lazy sequence out, `for` should already do that.

File: src/metabase/driver/sql_jdbc/execute.clj
Comment (walterl) [inline src/metabase/driver/sql_jdbc/execute.clj]: Can't this function be factored out? It'll make `run-query` easier to read.

File: src/metabase/driver.clj
Comment (walterl) [inline src/metabase/driver.clj]: This is expected to run for all non-SQL databases, right? I had a hunch, and then confirmed that it doesn't work with MongoDB (see exception below).

I expect the same is true for other non-SQL database, although I haven't confirmed that.

I also haven't dug into _why_ this happened. @camsaul do you have any ideas?

```
01-14 00:08:54 DEBUG middleware.log :: POST /api/dataset 200 [ASYNC: completed] 113.0 ms (16 DB calls) Jetty threads: 4/50 (2 idle, 0 queued) (63 total active threads) Queries in flight: 0
01-14 00:09:02 WARN middleware.process-userland-query :: Query failure {:status :failed,
 :class java.lang.Exception,
 :error
 "Query processor error: mismatched number of columns in query and results. Expected 11 fields, got 0\nExpected: [\"_id\" \"id\" \"name\" \"adminregion\" \"capitalCity\" \"incomeLevel\" \"iso2Code\" \"latitude\" \"lendingType\" \"long
itude\" \"region\"]\nActual: []",
 :stacktrace
 ("--> query_processor.middleware.annotate$check_correct_number_of_columns_returned.invokeStatic(annotate.clj:371)"
  "query_processor.middleware.annotate$check_correct_number_of_columns_returned.invoke(annotate.clj:364)"
  "query_processor.middleware.annotate$eval56789$fn__56791.invoke(annotate.clj:437)"
  "query_processor.middleware.annotate$eval56830$add_column_info_STAR___56835$fn__56839.invoke(annotate.clj:475)"
  "query_processor.middleware.annotate$eval56830$add_column_info_STAR___56835.invoke(annotate.clj:470)"
  "query_processor.middleware.annotate$add_column_info$fn__56852.invoke(annotate.clj:485)"
  "query_processor.middleware.format_rows$format_rows$fn__67384.invoke(format_rows.clj:82)"
  "query_processor$eval71850$execute_query__71855$fn__71856$fn__71857.invoke(query_processor.clj:74)"
  "driver$eval36832$fn__36833.invoke(driver.clj:340)"
  "query_processor$eval71850$execute_query__71855$fn__71856.invoke(query_processor.clj:71)"
  "query_processor$eval71850$execute_query__71855.invoke(query_processor.clj:64)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__64631.invoke(mbql_to_native.clj:40)"
  "query_processor.middleware.annotate$result_rows_maps__GT_vectors$fn__56946.invoke(annotate.clj:541)"
  "query_processor.middleware.annotate$add_column_info$fn__56852.invoke(annotate.clj:485)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__66722.invoke(cumulative_aggregations.clj:57)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__68529.invoke(resolve_joins.clj:184)"
  "query_processor.middleware.limit$limit$fn__67404.invoke(limit.clj:19)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__71650.invoke(results_metadata.clj:87)"
  "query_processor.middleware.format_rows$format_rows$fn__67384.invoke(format_rows.clj:82)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__65051.invoke(add_dimension_projections.clj:234)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__65718.invoke(add_source_metadata.clj:107)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__68587.invoke(resolve_source_table.clj:46)"
  "query_processor.middleware.add_row_count_and_status$add_row_count_and_status$fn__65563.invoke(add_row_count_and_status.clj:16)"
  "driver.mongo$eval101959$fn__101960$fn__101962$f__100870__auto____101964.invoke(mongo.clj:86)"
  "driver.mongo.util$_with_mongo_connection$fn__100861.invoke(util.clj:205)"
  "util.ssh$with_ssh_tunnel_STAR_.invokeStatic(ssh.clj:106)"
  "util.ssh$with_ssh_tunnel_STAR_.invoke(ssh.clj:91)"
  "driver.mongo.util$_with_mongo_connection.invokeStatic(util.clj:199)"
  "driver.mongo.util$_with_mongo_connection.invoke(util.clj:194)"
  "driver.mongo$eval101959$fn__101960$fn__101962.invoke(mongo.clj:85)"
  "query_processor.middleware.driver_specific$process_query_in_context$fn__66819.invoke(driver_specific.clj:12)"
  "query_processor.middleware.resolve_driver$resolve_driver$fn__68177.invoke(resolve_driver.clj:22)"
  "query_processor.middleware.store$initialize_store$fn__71689$fn__71690.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
  "query_processor.store$do_with_store.invoke(store.clj:40)"
  "query_processor.middleware.store$initialize_store$fn__71689.invoke(store.clj:10)"
  "query_processor.middleware.async$async__GT_sync$fn__64538.invoke(async.clj:23)"
  "query_processor.middleware.async_wait$runnable$fn__65785.invoke(async_wait.clj:89)"),
 :query
 {:type "query",
  :query {:source-table 76},
  :parameters [],
  :async? true,
  :middleware {:userland-query? true, :skip-results-metadata? true},
  :data-fn #function[metabase.api.dataset/eval72146/as-format-async--72151/fn--72152/fn--72154],
  :info
  {:executed-by 1,
   :context :csv-download,
   :query-hash [33, 103, -29, -79, -26, -51, 119, -34, 115, -111, 91, 44, -20, -90, -36, 54, -113, 63, 39, -64, -100, -93, 70, 101, 23, 12, -66, 73, 101, -28, 29, 124]}},
 :preprocessed
 {:type :query,
  :middleware {:userland-query? true, :skip-results-metadata? true},
  :data-fn #function[metabase.api.dataset/eval72146/as-format-async--72151/fn--72152/fn--72154],
  :info
  {:executed-by 1,
   :context :csv-download,
   :query-hash [33, 103, -29, -79, -26, -51, 119, -34, 115, -111, 91, 44, -20, -90, -36, 54, -113, 63, 39, -64, -100, -93, 70, 101, 23, 12, -66, 73, 101, -28, 29, 124]},
  :preprocessing-level 1,
  :database 27,
  :query
  {:source-table 76,
   :fields [[:field-id 134] [:field-id 135] [:field-id 138] [:field-id 141] [:field-id 140] [:field-id 142] [:field-id 133] [:field-id 137] [:field-id 143] [:field-id 136] [:field-id 139]],
   :limit 1048576}},
 :native
 {:projections (:_id :id :name :adminregion :capitalCity :incomeLevel :iso2Code :latitude :lendingType :longitude :region),
  :query
  [{"$project"
    {"_id" "$_id",
     "id" "$id",
     "name" "$name",
     "adminregion" "$adminregion",
     "capitalCity" "$capitalCity",
     "incomeLevel" "$incomeLevel",
     "iso2Code" "$iso2Code",
     "latitude" "$latitude",
     "lendingType" "$lendingType",
     "longitude" "$longitude",
     "region" "$region"}}
   {"$project"
    {"_id" "$_id",
     "id" "$id",
     "name" "$name",
     "adminregion" "$adminregion",
     "capitalCity" "$capitalCity",
     "incomeLevel" "$incomeLevel",
     "iso2Code" "$iso2Code",
     "latitude" "$latitude",
     "lendingType" "$lendingType",
     "longitude" "$longitude",
     "region" "$region"}}
   {"$limit" 1048576}],
  :collection "countries",
  :mbql? true}}

01-14 00:09:02 DEBUG middleware.log :: POST /api/dataset/csv 400 259.1 ms (35 DB calls)
"Query processor error: mismatched number of columns in query and results. Expected 11 fields, got 0\nExpected: [\"_id\" \"id\" \"name\" \"adminregion\" \"capitalCity\" \"incomeLevel\" \"iso2Code\" \"latitude\" \"lendingType\" \"longi
tude\" \"region\"]\nActual: []"
```

Test data imported with:

```
curl https://api.worldbank.org/countries\?per_page\=304\&format\=json | jq '.[1]' > countries.json
mongoimport -d metabase_11554_test -c countries --jsonArray < countries.json
```

**Review by walterl — CHANGES_REQUESTED** [review]
Mostly style changes, and then one biggy: this appears to break export for MongoDB and, I suspect, other non-SQL databases. See the relevant comment for more details.

Great work getting it this far, though!

Thanks for your contribution, and all your vim plugins that helped me do this review. :slightly_smiling_face:

File: src/metabase/driver.clj
Comment (tpope) [reply src/metabase/driver.clj]: This is directly related to the ad hoc middleware invocation discussed in https://github.com/metabase/metabase/pull/11554#discussion_r366109790. Adding the `annotate/result-rows-maps->vectors` middleware fixes it.

File: src/metabase/query_processor.clj
Comment (tpope) [reply src/metabase/query_processor.clj]: This is the one wart I couldn't figure out how to get rid of. I think ideally, `f` would receive the same fully decorated result that one would get when no callback is used at all. This poses 2 problems:

1. I don't know how to do this, short of fully rewriting the middleware stack.
2. At least one middleware - the one that adds the row count - would result in the entire result set being loaded into memory, defeating the purpose.

I settled on making ad hoc calls to the middleware necessary for export. The alternative, which I believe you are alluding to, would be to give up and hand the raw results directly to `f`, which could then invoke the middleware.  Calling into query processor middleware internals from export code felt more wrong to me than providing an incomplete set of middleware to all consumers. I can certainly flip it around if requested, but I hope there's a third option I'm not seeing.

File: src/metabase/query_processor/middleware/format_rows.clj
Comment (tpope) [reply src/metabase/query_processor/middleware/format_rows.clj]: Good catch, this snuck in from the original PR and is no longer necessary.

File: src/metabase/query_processor.clj
Comment (tpope) [reply src/metabase/query_processor.clj]: That's a good justification, but really it's a holdover from the original PR that had a separate `ostream` parameter. Will change if that's your decision.

File: src/metabase/driver/sql_jdbc/execute.clj
Comment (tpope) [reply src/metabase/driver/sql_jdbc/execute.clj]: Will circle back on this one when we have a game plan for middleware.

File: src/metabase/driver/sql_jdbc/execute.clj
Comment (lins05) [inline src/metabase/driver/sql_jdbc/execute.clj]: What about adding a comment to make it clear that this function preserves the laziness of the rows?

Comment (walterl) [reply src/metabase/driver/sql_jdbc/execute.clj]: @lins05 With the way Clojure handles laziness, I'm not sure that's necessary. I think it's more important to make it clear when lazy seqs are _realized_, rather than where they're preserved. Also, in the case where `:rows` is `[]`, such a comment could be misleading.

File: src/metabase/query_processor.clj
Comment (walterl) [reply src/metabase/query_processor.clj]: Ah. I didn't look at the original PR.

I'm not sure which way to go, really. On the one hand a distinct signature makes it clear which shape of the function you're using (and what return value to expect), on the other hand it's smelly when a change requires a bunch of changes in a bunch of different files. Then again, if the arguments _and_ the return value are different, maybe it should just be a separate function.

@camsaul What would you recommend?

File: src/metabase/query_processor.clj
Comment (walterl) [reply src/metabase/query_processor.clj]: > Calling into query processor middleware internals from export code felt more wrong to me...

Agreed! Maybe this calls for a different pipeline to be defined, for using in async/export cases like these?

I think I'll defer to @camsaul's knowledge of the intricacies of the query processor.

File: src/metabase/driver/sql_jdbc/execute.clj
Comment (tpope) [reply src/metabase/driver/sql_jdbc/execute.clj]: Personally I'm on team @walterl here. I was surprised to find that it wasn't already lazy.

File: src/metabase/driver.clj
Comment (camsaul) [inline src/metabase/driver.clj]: what happens if we run into an Exception when executing the query?

**Review by camsaul — CHANGES_REQUESTED** [review]
I think overall this change is a step in the right direction, but I'm worried that adding a new pattern for handling async queries is making the code too complicated, and harder to debug.

In about half of the query process code we're using an async pattern that looks like this:

```clj
;; a QP middleware fn
(defn an-async-middleware-fn [qp]
  ;; You can listen to the core.async `canceled-chan` to find out if the query is canceled
  (fn [query respond raise canceled-chan]
    (qp 
     (do-preprocessing query)
     (fn [results] (respond (do-postprocessing results))
     raise
     canceled-chan))
```

Since async code like this is so much harder to debug and write correctly, I haven't converted the entire query processor over to using the pattern. About half still uses a synchronous pattern:

```clj
(defn a-sync-middleware-fn [qp]
  (fn [query]
    (do-postprocessing (qp (do-preprocessing query)))))
```  

It looks to me like this PR is introducing another async pattern for use in things that are using the exist synchronous pattern. I would rather there not be two different patterns in use, let alone three. 

I think a better long-term fix would be to:

*  convert everything to using our existing async pattern, which adequately handles exceptions in async code as well as query cancellations
*  rework drivers to stream results for all queries (not just exported ones). It might make sense to switch to using a `core.async` channel for the output, or perhaps have just the `:rows` portion be a `core.async` channel

That being said, I don't want to let the perfect be the enemy of the good, so I can see us possibly merging it and using it as a starting point for making everything streaming in the future. These are my main concerns (please help me understand this PR better):

*  How does this handle query cancellations?
   *  If a query API request is canceled after receiving some (but not all) results, does the driver stop fetching rows from the Database?
*  What if a driver runs into an Exception when running a query, and never calls `callback` in `execute-query-callback`?
  *  What if the driver calls `callback` after some delay, say a minute?
*  How does `execute-query-callback` handle Exceptions?

I also think we'd want some tests to make sure this is actually streaming stuff in a lazy way. Here's one way I think you could do it: you could create a mock driver that implements `execute-query-callback` and maybe a few other methods, and returns some lazy sequence of results. It can have some way of checking whether the lazy seq was fully realized, e.g. by putting a value in an atom somewhere as a side-effect if it gets to the end. Then run a query against it and cancel before getting to the end, and check to make sure the lazy seq was never fully realized. You might also be able to use `with-redefs` and swap out some of the functions used to run SQL queries 

I want to make sure this actually implements streaming, it's not really a huge improvement if the output is streamed but the entire result set is realized in memory anyway.

**Comment by camsaul** [comment]
I hope those questions/requests are reasonable @tpope, once this is merged the core team will be responsible for maintaining it and working with it in the future and I want to make sure this isn't going to make the code harder to work with or introduce unforeseen consequences

File: src/metabase/driver.clj
Comment (tpope) [reply src/metabase/driver.clj]: For both the `sql-jdbc` and `:default` implementations, an exception is thrown as usual, and the callback is never invoked.

**Comment by tpope** [comment]
> I hope those questions/requests are reasonable @tpope, once this is merged the core team will be responsible for maintaining it and working with it in the future and I want to make sure this isn't going to make the code harder to work with or introduce unforeseen consequences

Say no more, this isn't my first rodeo. Unfortunately I'm out of time to allocate to this. If tests are required, then I'm going to be crossing my fingers that someone else can step up. I've answered your questions below as best as I can, and I've resolved conflicts and cleaned up the history to give anyone else who wants to take a shot at this a clean starting point.

> * How does this handle query cancellations?
>   
>   * If a query API request is canceled after receiving some (but not all) results, does the driver stop fetching rows from the Database?

If you cancel the export download, the CPU usage immediately drops to zero, which I am taking to mean this is performing correctly.

> * What if a driver runs into an Exception when running a query, and never calls `callback` in `execute-query-callback`?

If you asking in the context of an export web request, there is [a test that verifies an error is sent to the client](https://github.com/metabase/metabase/blob/2807015bf70e1b686badf5ecd55a8865e7134fb8/test/metabase/api/embed_test.clj#L433-L442). I put a lot of effort into getting this to work, and as a result I'm *pretty* sure it's doing the right thing.

> * What if the driver calls `callback` after some delay, say a minute?

The response begins streaming when `callback` is invoked. So on a provider like Heroku with a 30 second window to begin streaming, the request will still timeout. If we preemptively started the response before the callback was invoked, we could trigger Heroku's 55 second rolling window, but that would mean sending a `200 OK` before we're sure that things are, in fact, OK.

> * How does `execute-query-callback` handle Exceptions?

It allows them to bubble up, which seems in line with the behavior for `execute-query`.

**Comment by camsaul** [comment]
Superseded by #11832, which streams all QP results everywhere.

**Pull request closed without merging**
